Omit extraneous metadata copy in DrmBuffer

MicheleL-Intel opened the pull request:
DRM buffer details such as width, height etc are already present
  in the image metadata.  Store a reference of the passed-in HwcBuffer
  object with the exception of the shorthand buffer aliases (such as
  pitches_[] and offsets_[] which are useful for readibility.

  Jira: GSE-1509
  Test: Tested on KabyLake NUC

  Signed-off-by: Michele Lim <michele.lim@intel.com>

strassek commented on wsi/drm/drmbuffer.cpp:
```diff
@@ -77,23 +77,43 @@ DrmBuffer::~DrmBuffer() {
+  width_ = GetWidth();
+  height_ = GetHeight();
+
+  format_ = GetFormat();
```
I think we talked about this off-line, but I can't remember the reasoning behind keeping some of these copies around. Would it be possible to drop these and use only the image_.handle_->meta_data_ copies?

strassek commented on wsi/drm/drmbuffer.cpp:
```diff
@@ -77,23 +77,43 @@ DrmBuffer::~DrmBuffer() {
+
 void DrmBuffer::Initialize(const HwcBuffer& bo) {
   width_ = bo.width_;
   height_ = bo.height_;
```
Looks like these 2 assignments can be dropped.

MicheleL-Intel commented:
Thanks Kevin for your comments.  

Based on my recollection, we discussed this fix as being dual-phased.  First phase is to get rid of the pitches_[], offsets_[], and gem_handles_[] buffer writes which this PR resolves.  

Other than the fact that we want to eliminate unnecessary copies, every time we instantiate DrmBuffer, it creates the class member buffer objects which can be a synchronization issue when or if the data does change.  As far as I can tell, the buffers are set only once during  DrmBuffer::Initialize()

Not sure what others call these but I refer to them as "shorthand aliases".  I believe they were created for readability.

The second phase is to introduce macros with token concatenation, which will eliminate these shorthand aliases altogether.

Example define and usage (which will be incorporated in Phase 2)
# define METADATA (PARAM)     {image_.handle_->meta_data_.##PARAM}
METADATA (width_)
METADATA (height_)
METADATA (pitches_)
METADATA (offsets_)

strassek commented:
Understood, for consistency I'd like to see the remaining copies (width_, height_, etc..) removed, and use the long-form image_.handle... in their place.

MicheleL-Intel commented:
I will revise as suggested,  Thanks again.

MicheleL-Intel commented:
Removed the shorthand aliases, however, I'm having build issues with latest and errors appear to be unrelated to the changes.  Cleaned and currently rebuilding.

kalyankondapally commented on wsi/drm/drmbuffer.cpp:
```diff
@@ -77,29 +77,36 @@ DrmBuffer::~DrmBuffer() {
-  width_ = bo.width_;
-  height_ = bo.height_;
+void DrmBuffer::SaveMetaData(const HwcBuffer& bo) {
+  image_.handle_->meta_data_.width_ = bo.width_;
```
hmm do we really need to do this ?

in Initialize(image_.handle_->meta_data_);  it's the same metadata struct right ? we can avoid this whole copy. ?

MicheleL-Intel replied:
We discussed this briefly and I thought you said the image_.handle_.metadata doesn't get populated during init but until it's imported.  I will remove SaveMetaData.

harishkrupo commented:
@MicheleL-Intel Please rebase on top of master, travis issues are now fixed.

MicheleL-Intel commented:
Rebased, thanks.

kalyankondapally commented on wsi/drm/drmbuffer.cpp:
```diff
@@ -78,28 +78,12 @@ DrmBuffer::~DrmBuffer() {
-  }
-
-  format_ = bo.format_;
-  if (format_ == DRM_FORMAT_NV12_Y_TILED_INTEL || format_ == DRM_FORMAT_NV21)
```
You removed these checks ?

MicheleL-Intel replied:
These checks are for setting format_ which is no longer needed since the shorthand aliases are being removed.

kalyankondapally replied:
We still need this as these are custom formats on Android which we translate to standard DRM enums.

MicheleL-Intel replied:
Is it just format_ that needs to be preserved?

kalyankondapally replied:
I think u forgot to update the PR ?

MicheleL-Intel replied:
Hmmm, I thought I did so.  Let me check tomorrow.

kalyankondapally commented on wsi/drm/drmbuffer.cpp:
```diff
@@ -337,20 +327,20 @@ const MediaResourceHandle& DrmBuffer::GetMediaResource(MediaDisplay display,
   VASurfaceAttribExternalBuffers external;
   memset(&external, 0, sizeof(external));
-  uint32_t rt_format = DrmFormatToRTFormat(format_);
+  uint32_t rt_format = DrmFormatToRTFormat(image_.handle_->meta_data_.format_);
```
should be just format_ here and all other places which used format_ ? or wondering if we can store format value directly in meta_data_. did u check that and saw any failures ?

MicheleL-Intel replied:
You're right, this should be format_, I didn't see any failures when I tested this.  I'll check to see if format_ is being referenced elsewhere just in case.

js0701 commented:
@MicheleL-Intel  I like the idea of Phase2. How about do it and merge to this PR? Currently this PR make the code a little verbose

MicheleL-Intel commented:
Alright I'll submit a new patch with the macros.

MicheleL-Intel commented:
Replaced by #417
